doc/sphinxext/missing_references.py

```diff
@@ -0,0 +1,241 @@
+"""
+This is a sphinx extension to freeze your broken reference problems
+when using ``nitpicky = True``.
+
+The basic operation is:
+
+1. Add this extension to your ``conf.py`` extensions.
+2. Add ``missing_references_write_json = True`` to your ``conf.py``
+3. Run sphinx-build. It will generate ``missing-references.json``
+    next to your ``conf.py``.
+4. Remove ``missing_references_write_json = True`` from your
+    ``conf.py`` (or set it to ``False``)
+5. Run sphinx-build again, and ``nitpick_ignore`` will
+    contain all of the previously failed references.
+
+"""
+
+from collections import defaultdict
+import json
+import logging
+import os.path
+import posixpath
+
+from docutils.utils import get_source_line
+from docutils import nodes
+from sphinx.util import logging as sphinx_logging
+
+logger = sphinx_logging.getLogger(__name__)
+
+
+class MissingReferenceFilter(logging.Filter):
+    """
+    A logging filter designed to record missing reference warning messages
+    for use by this extension
+    """
+    def __init__(self, app):
+        self.app = app
+        super().__init__()
+
+    def _record_reference(self, record):
+        if not (getattr(record, 'type', '') == 'ref' and
+                isinstance(getattr(record, 'location', None), nodes.Node)):
+            return
+
+        if not hasattr(self.app.env, "missing_reference_warnings"):
+            self.app.env.missing_reference_warnings = defaultdict(set)
+
+        record_missing_reference(self.app,
+            self.app.env.missing_reference_warnings,
+            record.location)
+
+    def filter(self, record):
+        self._record_reference(record)
+        return True
+
+
+def record_missing_reference(app, record, node):
+    domain = node["refdomain"]
+    typ = node["reftype"]
+    target = node["reftarget"]
+    location = get_location(node, app)
+
+    dtype = "{}:{}".format(domain, typ)
```

Comment:
`dtype` means something in NumPy, so calling it this was a bit confusing at first.